dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/redis/CacheInvalidationListener.java

```diff
@@ -158,9 +158,18 @@ private Serializable getEntityId(String message) throws ClassNotFoundException {
       return getTrackedEntityAttributeValueId(idPart);
     } else if (CompleteDataSetRegistration.class.isAssignableFrom(entityClass)) {
       return getCompleteDataSetRegistrationId(idPart);
-    } else if (DataStatisticsEvent.class.isAssignableFrom(entityClass)) {
-      return Integer.parseInt(idPart);
     } else {
+      try {
+        // Best effort to try to identify classes with int IDs.
+        Field idField = entityClass.getDeclaredField("id");
+        Class<?> idType = idField.getType();
+        if (idType == int.class) {
+          return Integer.parseInt(idPart);
+        }
+      } catch (NoSuchFieldException e) {
+        // Ignore this exception, as it's expected unless it's a class with an int ID.
+      }
+      // In most cases the ID will be a long.
       return Long.parseLong(idPart);
```

Comment:
It looks like if there is no id field and the NoSuchFieldException is thrown that we still try parse the `idPart` as a long. Is that what we want to try here?
Some tests for this logic would be beneficial I think with some many conditional branches in this method.

Reply:
My thinking here is that if there is an ID field on the base class it will be an Integer. If you do a search for "int id;" you will see the result is only classes that has the ID field on the base class. So when it's not an int field, it will be a Long, mostly via the BaseIdentifiableObject. Making logic to catch every single case here is not the intention.